massa-consensus-worker/src/tests/inter_cycle_batch_finalization.rs

```diff
@@ -0,0 +1,158 @@
+/// the final roll registry was assumed to be attached to the last final block known by the graph,
+/// but that was inaccurate becaue PoS was the one holding the final roll registry,
+/// and PoS was not yet aware of the blocks that finalized during recursion,
+/// so it was actually still attached to G when B4 was checked.
+///
+/// The correction involved taking the point of view of PoS on where the final roll registry is attached.
+/// This test ensures non-regression by making sure B4 is propagated when B1 is received.
+#[tokio::test]
+#[serial]
+async fn test_inter_cycle_batch_finalization() {
+    let t0: MassaTime = 1000.into();
+    let staking_key = generate_random_private_key();
+    let creator_public_key = derive_public_key(&staking_key);
+    let creator_addr = Address::from_public_key(&creator_public_key);
+    let roll_price = Amount::from_str("42").unwrap();
+    let initial_ledger = vec![(
+        creator_addr,
+        LedgerData {
+            balance: roll_price, // allows the address to buy 1 roll
+        },
+    )]
+    .into_iter()
+    .collect();
+    let warmup_time: MassaTime = 1000.into();
+    let margin_time: MassaTime = 300.into();
+    let cfg = ConsensusConfig {
+        periods_per_cycle: 2,
+        delta_f0: 1,
+        thread_count: 1,
+        endorsement_count: 1,
+        max_future_processing_blocks: 10,
+        max_dependency_blocks: 10,
+        future_block_processing_max_periods: 10,
+        roll_price,
+        t0,
+        genesis_timestamp: MassaTime::now().unwrap().saturating_add(warmup_time),
+        ..ConsensusConfig::default_with_staking_keys_and_ledger(&vec![staking_key], &initial_ledger)
+    };
+
+    consensus_without_pool_test(
+        cfg.clone(),
+        async move |mut protocol_controller, consensus_command_sender, consensus_event_receiver| {
+            // wait for consensus warmup time
+            tokio::time::sleep(warmup_time.to_duration()).await;
+
+            let genesis_blocks: Vec<BlockId> = consensus_command_sender
+                .get_block_graph_status(None, None)
+                .await
+                .expect("could not get block graph status")
+                .best_parents
+                .iter()
+                .map(|(b, _p)| *b)
+                .collect();
+
+            // create B1 but DO NOT SEND IT
+            tokio::time::sleep(t0.to_duration()).await;
+            let (b1_id, b1_block, _) =
+                create_block(&cfg, Slot::new(1, 0), genesis_blocks.clone(), staking_key);
+
+            // create and send B2
+            tokio::time::sleep(t0.to_duration()).await;
+            let (b2_id, b2_block, _) = create_block_with_operations_and_endorsements(
+                &cfg,
+                Slot::new(2, 0),
+                &vec![b1_id],
+                staking_key,
+                vec![],
+                vec![create_endorsement(staking_key, Slot::new(1, 0), b1_id, 0)],
```

Comment:
Are the endowment need for this test to be a regression test ?

Reply:
the endorsement allows having delta_f0 = 1 (the minimum) while ensuring that every new block finalizes the previous one (by adding a fitness of 2 instead of 1)